pkg/controller/cluster_pools_pull_secret_provider/cluster_pools_pull_secret_provider.go

```diff
@@ -0,0 +1,217 @@
+
+func AddToManager(manager manager.Manager,
+	sourcePullSecretNamespace, sourcePullSecretName string) error {
+	log := logrus.WithField("controller", ControllerName)
+
+	client := manager.GetClient()
+	r := &reconciler{
+		log:                       log,
+		client:                    client,
+		sourcePullSecretNamespace: sourcePullSecretNamespace,
+		sourcePullSecretName:      sourcePullSecretName,
+	}
+	c, err := controller.New(ControllerName, manager, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	if err := c.Watch(
+		source.NewKindWithCache(&hivev1.ClusterPool{}, manager.GetCache()),
+		clusterPoolHandler(sourcePullSecretNamespace),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for clusterpools: %w", err)
+	}
+
+	if err := c.Watch(
+		source.NewKindWithCache(&corev1.Secret{}, manager.GetCache()),
+		secretHandler(sourcePullSecretNamespace, sourcePullSecretName, client),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for secrets: %w", err)
+	}
+
+	r.log.Info("Successfully added reconciler to manager")
+	return nil
+}
+
+func clusterPoolHandler(sourcePullSecretNamespace string) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(o ctrlruntimeclient.Object) []reconcile.Request {
+		clusterPool, ok := o.(*hivev1.ClusterPool)
+		if !ok {
+			logrus.WithField("type", fmt.Sprintf("%T", o)).Error("Got object that was not a ClusterPool")
+			return nil
+		}
+		if clusterPool.Namespace == sourcePullSecretNamespace {
+			return nil
+		}
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Namespace: clusterPool.Namespace,
+			Name:      clusterPool.Name,
+		}}}
+	})
+}
+
+func referPullSecret(clusterPool hivev1.ClusterPool, sourcePullSecretName string) bool {
+	if clusterPool.Spec.PullSecretRef == nil {
+		return false
+	}
+	if clusterPool.Spec.PullSecretRef.Name != sourcePullSecretName {
+		return false
+	}
+	return true
+}
+
+func requestsFactoryForSecretEvent(namespace string, client ctrlruntimeclient.Client, sourcePullSecretName string) []reconcile.Request {
+	clusterPools := &hivev1.ClusterPoolList{}
+	if err := client.List(context.TODO(), clusterPools, ctrlruntimeclient.InNamespace(namespace)); err != nil {
+		logrus.WithField("namespace", namespace).WithError(err).Error("Failed to list the cluster pools")
+		return nil
+	}
+
+	if namespace != "" {
+		for _, pool := range clusterPools.Items {
+			if referPullSecret(pool, sourcePullSecretName) {
+				return []reconcile.Request{{NamespacedName: types.NamespacedName{
+					Namespace: namespace,
+					Name:      pool.Name,
+				}}}
+			}
+		}
+		return nil
+	}
+	var requests []reconcile.Request
+	namespacesWithClusterPool := sets.NewString()
+	for _, pool := range clusterPools.Items {
+		if namespacesWithClusterPool.Has(pool.Namespace) || !referPullSecret(pool, sourcePullSecretName) {
```

Comment:
Same as above, please just enqueue all cluster pools and don't try to be too smart

Reply:
Updated.